[webkitbugspy] Implement ability to relate radars

briannafan opened the pull request:
#### d7bdf653cbd118b1081198813b8ea6c16bd6a4ae
<pre>
[webkitbugspy] Implement ability to relate radars
<a href="https://rdar.apple.com/120013230">rdar://120013230</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=266790">https://bugs.webkit.org/show_bug.cgi?id=266790</a>

Reviewed by Jonathan Bedard.

You can now create and modify relationships between radars using radarclient!
Added mocks for testing. Aligned bugzilla.relate functionality with radar.relate.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(RadarModel.__init__):
(RadarModel.relationships):
(RadarModel.add_relationship):
(Radar.Relationship):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker):
(Tracker.populate):
(Tracker.create_relationship):
(Tracker.relate):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: <a href="https://commits.webkit.org/273248@main">https://commits.webkit.org/273248@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd2e0c7f4c3a525906e29818ee3db6e0b9e6f2b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37389 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31326 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10609 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30284 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35232 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11440 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30906 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10003 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/35074 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10072 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30982 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38652 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31499 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31320 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36122 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10191 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8082 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34085 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12015 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7994 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10739 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11071 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
```diff
@@ -628,25 +628,10 @@ def test_relate_fail(self):
             self.assertEqual('\'fake_relation\' is an invalid relation', str(c.exception))
             self.assertEqual(issue.related, {'depends_on': [], 'blocks': [], 'regressions': [], 'regressed_by': []})
 
-    def test_relate_fail_type(self):
```
I don't understand why you're removing this test, it would seem like it's still valid?

briannafan replied:
True. I re-added it and updated the expected error message.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:
```diff
@@ -458,3 +460,37 @@ def test_set_keywords(self):
+            issue.relate(related_to=issue2)
+            self.assertEqual(issue.related, {'related-to': [issue2], 'blocked-by': [], 'blocking': [], 'parent-of': [], 'subtask-of': [], 'cause-of': [], 'caused-by': [], 'duplicate-of': [], 'original-of': []})
+
+    def test_relate(self):
```
This seems to be `test_parent_of`, right?
Would like to see a few more tests here, to test other relationship types. Doesn't have to be exhaustive, but two sees small.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
```diff
@@ -263,15 +264,28 @@ def commit_changes(self):
+        for r in related:
+            if not self.client.parent.issues[self.id].get('related'):
+                self.client.parent.issues[self.id]['related'] = set()
+            self.client.parent.issues[self.id]['related'].add(r)
```
The trouble with this approach is that it isn't encode-able. I would recommend saving as dictionaries like:
```
{
    'relationship': <type>,
    'id': <related issue id>,
}
```

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
```diff
@@ -442,7 +459,28 @@ def __init__(self, text=None, addedAt=None, addedBy=None):
+        TYPE_SUBTASK_OF = 'subtask-of'
+        TYPE_CAUSE_OF = 'cause-of'
+        TYPE_CAUSED_BY = 'caused-by'
+        TYPES = ['related-to', 'blocked-by', 'blocking', 'parent-of', 'subtask-of', 'cause-of', 'caused-by', 'duplicate-of', 'original-of']
```
This should probably contain the variable names instead of the raw strings.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
```diff
@@ -71,6 +71,8 @@ class Tracker(GenericTracker):
         TASK
     ]
 
+    RELATIONSHIP_TYPES = ['related-to', 'blocked-by', 'blocking', 'parent-of', 'subtask-of', 'cause-of', 'caused-by', 'duplicate-of', 'original-of']
```
Seeing these string duplicated for the 3rd time.

briannafan committed:
[webkitbugspy] Implement ability to relate radars
rdar://120013230
https://bugs.webkit.org/show_bug.cgi?id=266790

Reviewed by Jonathan Bedard.

You can now create and modify relationships between radars using radarclient!
Added mocks for testing. Aligned bugzilla.relate functionality with radar.relate.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/radar.py:
(RadarModel.__init__):
(RadarModel.relationships):
(RadarModel.add_relationship):
(Radar.Relationship):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker):
(Tracker.populate):
(Tracker.create_relationship):
(Tracker.relate):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: https://commits.webkit.org/273248@main
